proposals/0000-backpack.rst

```diff
@@ -0,0 +1,1284 @@
+
+    -- in Concat.hs
+    module Concat where
+    import Str
+    concat xs = foldr append null xs
+
+Locally defined ``hsig`` files are declared in the Cabal file via the
+``signatures`` field::
+
+    library concat-indef
+        signatures: Str
+        exposed-modules: Concat
+
+Signatures can also be inherited from other libraries (more on this
+shortly); we refer to the set of all locally defined and inherited
+signatures as the set of **required signatures**.
+A library with required signatures is called an **indefinite library**.
+As it is missing implementations for its required signatures, it cannot be
+compiled; however, it can still be type-checked (*separate
+type-checking*) and registered with the compiler, so that it can be used
+by other indefinite libraries which depend on it.  In contrast, a
+**definite library** is a library with no required signatures (any library that
+doesn't use Backpack features is a definite library).
+
+An indefinite library can be **instantiated** (possibly multiple times)
+with implementations for all of its required signatures, allowing it to
+be compiled.  Instantiation happens
+automatically when a user depends on an indefinite library and another
+library which provides modules with the same name as the signatures.
+For implementation reasons, it is only possible to fill required signatures with
+modules from ``build-depends`` (and not locally defined ones)::
+
+     library str-bytestring
+         exposed-modules: Str
+
+     library concat-bytestring
+         build-depends: str-bytestring, concat-indef
+         reexported-modules:
+             -- Concat from concat-indef is instantiated
+             -- with Str from str-bytestring.  We can
+             -- reexport it under a qualified name for
+             -- more convenient use.
+             Concat as Concat.ByteString
+
+Thus, indefinite libraries can be thought of parametrized modules,
+but rather than explicitly specifying each parameter, it is
+implicitly specified with module namespaces.  This process
+of determining the explicit instantiations is called **mix-in linking**.
+
+An indefinite library can be instantiated to various degrees.
+Compilation does not occur unless *all* required signatures are implemented,
+allowing a compiler can optimize as if Backpack was not present (*no
+performance overhead.*)  An indefinite library can also be partially
+instantiated, or not instantiated at all.  If a required signature is not
+instantiated, it gets inherited by the user of the library::
+
+    -- in the Cabal file
+    library stringutils-indef
+        -- No Str module in scope, so Str is left uninstantiated,
+        -- giving stringutils-indef an (implicit) requirement
+        -- on Str.
+        build-depends: concat-indef
+        exposed-modules: StringUtils
+
+    -- in StringUtils.hs
+    module StringUtils where
+    import Concat
+    import Str -- the signature is importable
```

Comment:
We played around with a few possible semantics, but what ended up being easiest in the end is that signatures are always visible (importable), until you instantiate them, then they go away, as you said. Visibility is the same thing as importability.